Add try_conv and try_cast

dhardy opened the pull request:
Closes #4. I see two potential uses for these additional conversions (though didn't yet need them myself):

1. If the conversion is fallible, and failure should be handled at run-time.
2. If the conversion may be fallible and we want to ensure a check occurs (without using the `always_assert` feature).

~~TODO: float conversions.~~ Code share with `conv` is minimal and probably that's for the best.

I'll just leave this here for now. If @BartMassey or @vks or anyone else wishes to comment, please do.

dhardy committed:
Add try_conv and try_cast

BartMassey commented:
Awesome. Will review tomorrow at latest.

vks commented on CHANGELOG.md:
```diff
@@ -1,6 +1,10 @@
 
+## [0.4.0] — 2021-04-??
+
+-   Add `try_conv` and `try_cast` methods
```
Didn't you also add libm support?

dhardy replied:
That was part of `no_std` support on the previous release. Guess I could have mentioned it.

vks commented on src/lib.rs:
```diff
@@ -35,6 +35,32 @@
 use core::mem::size_of;
 
+/// Error types for conversions
+#[derive(Clone, Debug, PartialEq, Eq)]
```
What about `Copy`?

dhardy replied:
After that thing we had with `ThreadRng` I won't make public APIs `Copy` without a good reason and here there isn't one.

BartMassey commented on src/lib.rs:
```diff
@@ -35,6 +35,32 @@
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "easy-cast conversion error: {}",
```
Would prefer just "cast conversion error" here, probably. Not a big deal either way.

BartMassey commented on src/lib.rs:
```diff
@@ -259,20 +364,38 @@ impl Conv<f64> for f32 {
-                    let src_digits = src_ty_bits - (x.leading_zeros() + x.trailing_zeros());
-                    let dst_digits = core::$y::MANTISSA_DIGITS;
-                    assert!(src_digits <= dst_digits);
+                    Self::try_conv(x).expect("inexact int→float conversion")
```
Would suggest "inexact int to float conversion" here for friendliness with non-Unicode environments.

BartMassey reviewed (commented):
Take a look at the small changes I suggested and take them or not. I didn't spot anything else in a quick check.

dhardy committed:
Address review: simplify error messages

dhardy committed:
Add try_conv_trunc

dhardy committed:
Add try_conv_nearest/try_conv_floor/try_conv_ceil

dhardy committed:
Add try_cast_trunc/nearest/floor/ceil

dhardy committed:
Revise doc

dhardy committed:
Replace assert_range and assert_non_neg with assert_int

dhardy committed:
Remove Conv<f64> for f32

This probably shouldn't be included, just like Conv<f32>
and Conv<f64> are not implemented for int target types.

dhardy committed:
Test all configurations

dhardy committed:
Update changelog

dhardy commented:
Updated (fallible float conversions, doc revision, a few other bits).

Should be ready for a new release tomorrow.

dhardy committed:
Update crate version

dhardy committed:
Set MSRV to 1.32.0

This is the oldest version supporting Edition 2018 and thus
the oldest which will work without further changes.

dhardy commented:
After a little testing: 1.32.0 is the oldest version supporting Edition 2018 and works fine for us, so it is now the MSRV.

BartMassey commented:
Thanks much as always!
